File: ListOperations.scala
val testList = List(1,2,3,4,5,6,7,8,9,10)
println(fold((x: Int, y: Int) => x + y,
            0,
            map((x : Int) => x * 2,
                filter((x : Int) => x > 5,
                    filter((x : Int) => x % 2 == 1, testList)
                )
            )
        ))

println(testList
                .filter(x => x > 5)
                .filter(x => x % 2 == 1)
                .map(x => x * 2)
                .reduce((x, y) => x + y))

println(testList
                .filter(isGT5)
                .filter(isOdd)
                .map(doubleIt)
                .headOption)
val x = 8

Stream
    .from(1)
    .map(x => power(x, x))
    .take(x)
    .foreach(println)

def power(x: BigInt, y : BigInt): BigInt = {
    if (y == 0)
        1
    else if (y.mod(BigInt(2)) == 0) {
        val m = power(x, y./(BigInt(2)))
        m.*(m)
    } else x.*(power(x, y.-(BigInt(1))))
}


def isOdd(x : Int) : Boolean = {
    println("isOdd called... %d".format(x))
    x % 2 == 1
}

def isGT5(x : Int) : Boolean = {
    println("isGT5 called... %d".format(x))
    x > 5
}

def doubleIt(x : Int) : Int = {
    println("doubleIt called... %d".format(x))
    x * 2
}

def map[A, B](f: (A) => B, list: List[A]) : List[B] = list match {
    case (x :: xs) => {
        f(x) +: map(f, xs)
    }
    case _ => Nil
}

def filter[A](f: (A) => Boolean, list: List[A]) : List[A] = list match {
    case (x :: xs) => {
        if(f(x))
            x +: filter(f, xs)
        else
            filter(f, xs)

    }
    case _ => Nil
}

def fold[A](f: (A, A) => A, acc: A, list: List[A]) : A = list match {
    case (x :: xs) => {
        fold(f, f(acc, x), xs)
    }
    case _ => acc
}

File: ListOperations.java
import java.util.function.*;
import java.util.*;
import java.util.stream.*;
import java.math.BigInteger;

class ListOperations {
    public static void main(String[] args){
        List<Integer> values = Arrays.asList(1,2,3,4,5,6,7,8,9,10);
        // System.out.println(fold((a, b) -> a + b,
        //                         0,
        //                         map(e -> e * 2,
        //                             filter(e -> e % 2 == 1,
        //                                 filter(e -> e > 5, values)
        //                             )
        //                         )
        //                     ));
        // System.out.println(values
        //                     .stream()
        //                     .filter(e -> e > 5)
        //                     .filter(e -> e % 2 == 1)
        //                     .map(e -> e * 2)
        //                     .reduce(0, (a,b) -> a + b));
        System.out.println(values
                            .stream()
                            .filter(ListOperations :: isGT5)
                            .filter(ListOperations :: isOdd)
                            .map(ListOperations :: doubleIt)
                            .findFirst());
        // int x = 8;
        // Stream
        //     .iterate(BigInteger.ONE, e -> e.add(BigInteger.ONE))
        //     .map(e -> power(e, e))
        //     .limit(x)
        //     .forEach(System.out::println);
    }

    static BigInteger power(BigInteger x, BigInteger y){
        if (y.equals(BigInteger.ZERO)) return BigInteger.ONE;
        else if (y.mod(new BigInteger("2")).equals(BigInteger.ZERO)) {
            BigInteger m  = power(x, y.divide(new BigInteger("2")));
            return m.multiply(m);
        } else return x.multiply(power(x, y.subtract(BigInteger.ONE)));
    }

    static int doubleIt(int x){
        System.out.println(String.format("doubleIt called... %d", x));
        return x * 2;
    }

    static boolean isOdd(int a){
        System.out.println(String.format("isOdd called... %d", a));
        return a % 2 == 1;
    }

    static boolean isGT5(int a){
        System.out.println(String.format("isGT5 called... %d", a));
        return a > 5;
    }

    static <T, R> List<R> map(Function<T, R> f, List<T> list){
        ArrayList<R> result = new ArrayList<>();
        if(!list.isEmpty()){
            T first = list.get(0);
            result.add(f.apply(first));
            result.addAll(map(f, list.subList(1, list.size()))); //returns boolean
            return result;
        }
        else return result;
    }

    static <T> List<T> filter(Function<T, Boolean> f, List<T> list){
        ArrayList<T> result = new ArrayList<>();
        if(!list.isEmpty()){
            T first = list.get(0);
            if(f.apply(first))
                result.add(first);
            result.addAll(filter(f, list.subList(1, list.size()))); //returns boolean
            return result;
        }
        else return result;
    }

    static <T> T fold(BinaryOperator<T> f, T accum, List<T> list){
        ArrayList<T> result = new ArrayList<>();
        if(!list.isEmpty()){
            T first = list.get(0);
            return fold(f, f.apply(accum, first), list.subList(1, list.size()));
        }
        else return accum;
    }

}
